status/controller.go

```diff
@@ -2,53 +2,158 @@ package status
 
+	ConditionCount.DeletePartialMatch(objectLabels)
 	if err := c.kubeClient.Get(ctx, req.NamespacedName, o); err != nil {
 		return reconcile.Result{}, client.IgnoreNotFound(err)
 	}
 
+	observedConditions := c.observedConditions[req]
+	c.observedConditions[req] = o.StatusConditions()
+
 	for _, condition := range o.GetConditions() {
-		ConditionCount.MustCurryWith(labels).With(prometheus.Labels{
-			metricLabelConditionType:   condition.Type,
-			metricLabelConditionStatus: string(condition.Status),
+		ConditionCount.MustCurryWith(objectLabels).With(prometheus.Labels{
+			MetricLabelConditionType:   string(condition.Type),
+			MetricLabelConditionStatus: string(condition.Status),
 		}).Set(1)
+		// Detect and record status transitions. This approach is best effort,
+		// since we may batch multiple writes within a single reconcile loop.
+		// It's exceedingly difficult to atomically track all changes to an
+		// object, since the Kubernetes is evenutally consistent by design.
+		// Despite this, we can catch the majority of transition by remembering
+		// what we saw last, and reporting observed changes.
+		//
+		// We rejected the alternative of tracking these changes within the
+		// condition library itself, since you cannot guarantee that a
+		// transition made in memory was successfully persisted.
+		//
+		// Automatic monitoring systems must assume that these observations are
+		// lossy, specifically for when a condition transition rapidly. However,
+		// for the common case, we want to alert when a transition took a long
+		// time, and our likelyhood of observing this is much higher.
+		if observedCondition := observedConditions.Get(condition.Type); observedCondition != nil && observedCondition.GetStatus() != condition.GetStatus() {
+			duration := condition.LastTransitionTime.Time.Sub(observedCondition.LastTransitionTime.Time).Seconds()
+			log.FromContext(ctx).Info("observed transition",
+				"object", object.GKNN(o),
+				"fromCondition", observedCondition,
+				"toCondition", condition,
+				"duration", duration,
+			)
+			event.FromContext(ctx).Publish(NewStatusConditionTransitionEvent(o, condition))
+			ConditionDuration.MustCurryWith(objectLabels).With(prometheus.Labels{
+				MetricLabelConditionType:   string(observedCondition.Type),
+				MetricLabelConditionStatus: string(observedCondition.Status),
+			}).Observe(float64(duration))
+		}
 	}
 	return reconcile.Result{}, nil
 }
+
+func NewStatusConditionTransitionEvent(object Object, condition Condition) event.Event {
+	return event.Event{
+		InvolvedObject: object,
+		Type:           string(lo.Ternary(condition.Severity == ConditionSeverityInfo && condition.Status == metav1.ConditionFalse, v1.EventTypeNormal, v1.EventTypeWarning)),
+		Reason:         string(condition.Type),
+		Message:        string(condition.Reason),
+	}
+}
+
+// Cardinality is limited to # objects * # conditions * # objectives
+var ConditionDuration = prometheus.NewSummaryVec(
```

Comment:
How useful is this summary vector if it isn't summarizing any information? There's a bit of a conversation that happens around using summary vectors when you are trying to roll-up information. I'm curious how this is going to work if we are trying to roll-up a bunch of P99 information into a top-level P99? How are we going to do that? Are we going to average across the different elements from the different namespaces, weight them in some way? 

Ref: https://prometheus.io/docs/practices/histograms/#:~:text=interval%20of%20%CF%86.-,Two%20rules%20of%20thumb,-%3A

Reply:
Maybe worth an offline discussion. I'm new to summary metrics, and may be missing something.